### PR TITLE
feat: add conventional commit formatting

### PR DESCRIPTION
**Pull request by ml019**
## Description
Add utility routines to format the message of a git commit according to the conventional commit (https://www.conventionalcommits.org/).

Add support for local directory control to the git utility routines.

Split the formatting of the git url from the other routines which use it.

## Motivation and Context
Permit the working directory to not be within the repo.

This PR lays the groundwork for removing the use of tags to record the execution of process steps in favour of more complete use of commits. It will also permit the replacement of manageRepo.sh with more direct git control in scripts via the utility routines.

## How Has This Been Tested?
Local testing

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] None

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] None of the above.

### CODE REVIEW

File: execution/utility.sh [hunk 1]
```diff
@@ -2267,6 +2289,49 @@ function git_rm() {
...
+function format_conventional_commit() {
+  local type="${1,,}"; shift
+  local scope="${1,,}"; shift
+  local description="${1}"; shift
+  local body="${1}"; shift
+  local footer="${1}"; shift
+  local breaking="${1}"; shift
+
+  local result=""
+  printf -v result "%s%s%s: %s\n" "${type}" "${scope:+(${scope})}" "${breaking:+!}" "${description}"
```
Comment (roleyfoley) [inline execution/utility.sh:2306]: Might be worth setting the header max length to 50 chars in line with the general git standards

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: execution/utility.sh (hunk 1 above)
Comment (ml019) [reply execution/utility.sh:2306]: Trouble is the plan was to put the list of deployment units as the description which will likely blow the 50 char limit. The list can be put in the body but is probably useful in the description.

Suggestions for alternate approach?

File: execution/utility.sh [hunk 2]
```diff
@@ -2171,26 +2172,46 @@ function manage_waf_logging() {
...
+function format_git_url() {
   local repo_provider="$1"; shift
   local repo_host="$1"; shift
   local repo_path="$1"; shift
+
+  if [[  (-n "${repo_provider}") &&
+      (-n "${repo_host}") &&
+      (-n "${repo_path}") ]]; then
+    local credentials_var="${repo_provider^^}_CREDENTIALS"
+    printf "https://%s@%s/%s" "${!credentials_var}" "${repo_host}" "${repo_path}"
```
Comment (RossMurr4y) [inline execution/utility.sh:2200]: Do we need something here to gracefully fail on `${!credentials_var}` ?

File: execution/utility.sh (hunk 1 above)
Comment (roleyfoley) [reply execution/utility.sh:2306]: Maybe just the job name plus the job number 
The details of the deployment can be included in the body of the message

Comment (ml019) [reply execution/utility.sh:2306]: Nice to be able to search first lines for units. Routines don't define content so I'll add the limit. Anything truncated should flow onto the first line of the details.

File: execution/utility.sh (hunk 2 above)
Comment (ml019) [reply execution/utility.sh]: I now treat this as meaning the git repo is public and no credentials are required. The inclusion of the ampersand separator is now conditional to ensure the url is correctly formatted.

File: execution/utility.sh (hunk 1 above)
Comment (ml019) [reply execution/utility.sh:2309]: Decided to just truncate and leave it to the consumer to worry about this, given it controls the body content. If preserving the truncated content is important, the consumer should manage the body content accordingly.

**Comment by ml019** [comment]
@roleyfoley @RossMurr4y feedback addressed. Ready for review and merge.

**Review by roleyfoley — APPROVED** [review]

**Pull request merged**